docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
-constraints for generic class type parameters.  For example, the type
+constraints for generic class type parameters. For example, the type
 parameter of the generic class ``typing.IO`` is constrained (only
 ``IO[str]``, ``IO[bytes]`` and ``IO[Any]`` are valid)::
 
   class IO(Generic[AnyStr]): ...
+
+
+Invalid overload definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should enforce the following rules for overload definitions.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+An overload implementation should be present for all overload function
+definitions. Type checkers should report an error or warning if an
+implementation is missing. Overload definitions within stub files,
+protocols, and abstract base classes are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Similarly, if one overload
+is ``async``, all overloads must be ``async`` as well as the implementation.
+Type checkers should report an error if these conditions are not met.
+
+If one or more overloads are decorated with ``@final`` but the
+implementation is not, an error should be reported.
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that its signature is consistent with all of its associated overload
+signatures. The implementation should accept all potential argument lists
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be assignable to the input signatures
+of all overloads, and the return type of all overloads should be assignable to
+the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+Overlapping overloads
+^^^^^^^^^^^^^^^^^^^^^
+
+If two overloads can accept the same set of arguments but have
+different return types, the overloads are said to "partially overlap".
+This condition is indicative of a programming error and should
+be reported by type checkers::
+
+  # These overloads partially overlap because both accept an
+  # argument of type ``Literal[0]``, but their return types
+  # differ.
+
+  @overload
+  def func(x: Literal[0]) -> int: ...
+  @overload
+  def func(x: int) -> str: ...
+
```

Comment:
Could we add an example with kwargs here? Stating that bar should be keyword only for the second signature?

```python
@overload
def foo(*, bar: None = None) -> None: ...  # Overloaded function signatures 1 and 2 overlap with incompatible return types
@overload
def foo(*, bar: bool = ...) -> bool: ...
def foo(*, bar: bool | None = None) -> bool | None: ...
```

Reply:
Sorry, I don't follow you. Are you saying that you think a type checker should emit an error in this example? I don't think that's the case based on the rules. And mypy and pyright agree with this, since neither of them emit an error in this case.

Reply:
Also see https://stackoverflow.com/questions/74901020/python-typing-mypy-errors-with-overload-overlap-when-signatures-are-different.

Reply:
And https://github.com/python/mypy/issues/6580.

Reply:
My goal here isn't to provide positive and negative examples that cover all possible cases. That wouldn't be feasible. The code sample shows one one illustrative example. The general rule applies in all cases.

Reply:
I agree that the above example should be flagged as an error. Neither mypy nor pyright do today. I'll fix this in the next release of pyright. We should also include some of these examples in the conformance tests if/when this spec change is accepted.